# Alumno


*María Isabel Pérez Lisón*
*Ana Luo Sánchez Botella*


# Ejercicios de nivel básico

## Listas y diccionarios

### Dada una frase sin signos de puntuación, calcula la lista de las longitudes de las palabras que contiene.

Resuélvelo tanto con un bucle clásico como usando *list comprehensions*.

In [14]:
sentence = "the quick brown fox jumps over the lazy dog"

In [18]:
# Solución con un bucle clásico
s = []

for i in sentence.split():
    s.append(len(i))

print(s)


[3, 5, 5, 3, 5, 4, 3, 4, 3]


In [21]:
# Solución con list comprehension
s=[len(a) for a in sentence.split()]

print(s)

[3, 5, 5, 3, 5, 4, 3, 4, 3]


### Dada una lista de números, obtén la lista de los elementos que son positivos.

Resuélvelo tanto con un bucle clásico como usando *list comprehensions*.

In [23]:
numbers = [2.2, -3.4, 1.0, 5.6, -1.2]

In [24]:
# Solución con un bucle clásico
l=[]
for x in numbers:
    if x>0:
        l.append(x)

print(l)


[2.2, 1.0, 5.6]


In [ ]:
# Solución con list comprehension
[x for x in numbers if x>0]

[2.2, 1.0, 5.6]

### Dado el siguiente diccionario con pares `"nombre": peso`, extrae una lista con los nombres de los vehículos que pesan menos de 5000 kilos.

Los nombres de la lista extraída deben estar en minúsculas.

Resuélvelo usando *list comprehensions*.

In [24]:
vehicles = {
    "Sedan": 1500,
    "SUV": 2000,
    "Pickup": 2500,
    "Minivan": 1600,
    "Van": 2400,
    "Semi": 13600,
    "Bicycle": 7,
    "Motorcycle": 110
}

In [43]:
# Solución con list comprehension
n = [x for x,y in vehicles.items() if y < 5000]

print(n)

['Sedan', 'SUV', 'Pickup', 'Minivan', 'Van', 'Bicycle', 'Motorcycle']


### Crea una lista `lst2` con el tipo de cada elemento de la lista `lst1`.

Resuélvelo tanto con un bucle clásico como usando *list comprehensions*.

In [2]:
lst1 = [3.14, 66, "osito", True, [], {}]

In [6]:
# Solución con bucle clásico
lst2=[]
for elemento in lst1:
    lst2.append(type(elemento))
print(lst2)


[<class 'float'>, <class 'int'>, <class 'str'>, <class 'bool'>, <class 'list'>, <class 'dict'>]


In [73]:
# Solución con list comprehension
lst2= [type(elemento) for elemento in lst1]
print(lst2)

[<class 'float'>, <class 'int'>, <class 'str'>, <class 'bool'>, <class 'list'>, <class 'dict'>]


### Usando la función `zip()`, crea un diccionario que forme sus pares `"clave": valor` a partir de las listas `lst1` y `lst2`.

In [44]:
lst1 = ["Mercurio", "Venus", "Tierra", "Marte"]
lst2 = [57.9, 108.2, 149.6, 227.9]

In [69]:
# Solución
dict(zip(lst1, lst2))

{'Mercurio': 57.9, 'Venus': 108.2, 'Tierra': 149.6, 'Marte': 227.9}

### Crea una lista ordenada de tuplas `("producto", precio)` a partir de las listas `productos` y `precios`.

Pista: Usa las funciones `zip()` y `sorted()`

In [46]:
productos = ["Teclado", "Ratón", "Monitor", "Altavoces"]
precios = [25.50, 15.99, 149.00, 45.00]

In [70]:
# Solución
d=list(zip(productos,precios))
print(sorted(d))

[('Altavoces', 45.0), ('Monitor', 149.0), ('Ratón', 15.99), ('Teclado', 25.5)]


### Crea un diccionario que asocie cada letra del rango a-z con su correspondiente codificación en Unicode. 

Pista: Usa las funciones `ord()` (que devuelve el código Unicode de un carácter) y `chr()` (que devuelve la letra asociada a su código Unicode): `{'a': 97, 'b': 98, 'c': 99, ..., 'z': 122}`.

Resuélvelo tanto con un bucle clásico como usando *dict comprehensions*.

In [2]:
# Solución con bucle clásico
import string
dic = {}

for letra in string.ascii_lowercase:
    dic[letra] = ord(letra)

print(dic)

{'a': 97, 'b': 98, 'c': 99, 'd': 100, 'e': 101, 'f': 102, 'g': 103, 'h': 104, 'i': 105, 'j': 106, 'k': 107, 'l': 108, 'm': 109, 'n': 110, 'o': 111, 'p': 112, 'q': 113, 'r': 114, 's': 115, 't': 116, 'u': 117, 'v': 118, 'w': 119, 'x': 120, 'y': 121, 'z': 122}


In [3]:
# Solución con dict comprehension
import string
dic = {letra : ord(letra) for letra in string.ascii_lowercase }

print(dic)

{'a': 97, 'b': 98, 'c': 99, 'd': 100, 'e': 101, 'f': 102, 'g': 103, 'h': 104, 'i': 105, 'j': 106, 'k': 107, 'l': 108, 'm': 109, 'n': 110, 'o': 111, 'p': 112, 'q': 113, 'r': 114, 's': 115, 't': 116, 'u': 117, 'v': 118, 'w': 119, 'x': 120, 'y': 121, 'z': 122}


### Genera un diccionario que contenga las letras del rango a-z cuya codificación en Unicode sea un número par usando *dict comprehensions*.

Puedes partir del diccionario del ejercicio anterior.

In [5]:
# Solución
import string
dict={letra for letra in string.ascii_lowercase if ord(letra) %2 ==0}
print(sorted(dict))



['b', 'd', 'f', 'h', 'j', 'l', 'n', 'p', 'r', 't', 'v', 'x', 'z']


### Convierte una lista que contiene listas (bidimensional) en una lista plana.

Resuélvelo con una *list comprehension*

In [6]:
listas = [[1, 2, 3],[4, 5, 6]]

In [18]:
# Solución: [1, 2, 3, 4, 5, 6]
lf=[x for sublista in listas for x in sublista]
print(lf)

[1, 2, 3, 4, 5, 6]


### Programa una función `list_is_even(numbers: List[int])`

Esta función debe recibir una lista de números enteros y devolver una lista de booleanos indicando si cada número de la lista original es par. Por ejemplo, `list_is_even([1, 2, 3, 4, 10])` devolvería `[False, True, False, True, True]`.

In [23]:
# Solución
from typing import List
def list_is_even(numbers: List[int]):
    final=[]
    for x in numbers:
        if x %2 ==0:
            final.append(True)
        else:
            final.append(False)
    return final

list_is_even([1,2,3,4,10])


[False, True, False, True, True]

## Cadenas

### Construye una cadena que tenga todos los nombres de los vehículos del diccionario `vehicles` separados por comas.

Pista: Usa la función `join()`

In [29]:
# Solución
vehicles = {
    "Sedan": 1500,
    "SUV": 2000,
    "Pickup": 2500,
    "Minivan": 1600,
    "Van": 2400,
    "Semi": 13600,
    "Bicycle": 7,
    "Motorcycle": 110
}
n = [x for x,y in vehicles.items()]
cadena= ', '.join(n)
print(cadena)

Sedan, SUV, Pickup, Minivan, Van, Semi, Bicycle, Motorcycle


### Genera una cadena con retornos de carro, de manera que se imprima como un poema, con la siguiente lista de frases.

Pista: Usa las funciónes `strip()` y `join()`

In [ ]:
frases = [
    '  En un lugar de la Mancha,  ',
    'de cuyo nombre no quiero acordarme,',
    'no ha mucho tiempo que vivía',
    'un hidalgo de los de lanza en astillero,',
    'adarga antigua, rocín flaco y galgo corredor.  '
]

In [47]:
# Solución
#poema=[n.rstrip() for n in frases]
#print(poema)
poema = '\n'.join(frase.strip() for frase in frases)
print(poema)
    

En un lugar de la Mancha,
de cuyo nombre no quiero acordarme,
no ha mucho tiempo que vivía
un hidalgo de los de lanza en astillero,
adarga antigua, rocín flaco y galgo corredor.


### Convierte una cadena con una fecha en formato 'AAAA-MM-DD' al formato 'DD/MM/AAAA'.

Pista: Usa las funciones `split()` y `join()`

In [48]:
# Pista: la siguiente línea te puede facilitar la tarea.

a, b, c = [1, 2, 3]
print(a,b,c)

1 2 3


In [58]:
# Solución
fecha_og = '2026-03-26'
fecha = fecha_og.split('-')
ordenado = [fecha[2], fecha[1], fecha[0]]
final = "/".join(ordenado)# Unimos la lista con un guion
print(final)

26/03/2026


### Convierte a mayúsculas el primer de carácter de cada palabra de una lista de cadenas de caracteres.

Utiliza como entrada la lista de cadenas de caracteres `frases` definida anteriormente. El resultado debe ser una única cadena: "En Un Lugar De La Mancha, De Cuyo Nombre No Quiero Acordarme, No Ha Mucho Tiempo Que Vivía Un Hidalgo De Los De Lanza En Astillero, Adarga Antigua, Rocín Flaco Y Galgo Corredor.".

Resuélvelo tanto con un bucle clásico como usando *list comprehensions*.

Pista: Usa la función `capitalize()`

In [88]:

frases = [
    '  En un lugar de la Mancha,  ',
    'de cuyo nombre no quiero acordarme,',
    'no ha mucho tiempo que vivía',
    'un hidalgo de los de lanza en astillero,',
    'adarga antigua, rocín flaco y galgo corredor.  '
]
# Solución con bucle clásico
final=[]    
for palabra in frases:
    lista = palabra.strip().split(' ')
    for x in lista:
        final.append(x.capitalize())
resultado=' '.join(final)
print(resultado)

En Un Lugar De La Mancha, De Cuyo Nombre No Quiero Acordarme, No Ha Mucho Tiempo Que Vivía Un Hidalgo De Los De Lanza En Astillero, Adarga Antigua, Rocín Flaco Y Galgo Corredor.


In [89]:
# Solución con list comprehension
cadena = " ".join(palabra.capitalize() for frase in frases
                for palabra in frase.strip().split())
print(cadena)
            

En Un Lugar De La Mancha, De Cuyo Nombre No Quiero Acordarme, No Ha Mucho Tiempo Que Vivía Un Hidalgo De Los De Lanza En Astillero, Adarga Antigua, Rocín Flaco Y Galgo Corredor.


# Ejercicios de nivel intermedio

## Listas, diccionarios, conjuntos, tuplas y cadenas

### Programa una función llamada `is_valid_zipcode(zipcode: str)`

Esta función recibe una cadena y devuelve un *booleano* indicando si la cadena es un código postal válido de la Región de Murcia. Consideraremos que una cadena es un código postal murciano válido si está compuesta de 5 dígitos y empieza por '30'. Por ejemplo, `is_valid_zip_code('30500')` devolvería `True`, mientras que `is_valid_zip_code('3050a')`, `is_valid_zip_code('3050')` y `is_valid_zip_code('43050')` devolverían `False`.

In [98]:
# Solución
def is_valid_zipcode(zipcode:str):
    caracteres= zipcode[:2]
    #print(caracteres)
    if caracteres=='30' and len(zipcode) ==5:
        return True
    return False

print(is_valid_zipcode('30500'))
print(is_valid_zipcode('3050a'))
print(is_valid_zipcode('3050'))
print(is_valid_zipcode('43050'))

True
True
False
False


### Programa una función llamada `primos(numero: int)` 

Esta función debe recibir un número y devolver una lista con los números primos que se encuentran entre 2 y ese número. Por ejemplo:

`primos(50)`

devolvería la siguiente lista:

`[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]`

In [107]:
# Solución
def primos(numero:int):
    lista=[]

    for i in range (2, numero+1):
        for divisor in range(2,i):
            if i % divisor == 0:
                break
        else:
            lista.append(i)
    return lista
primos(50)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]

### Programa una función llamada `word_search(strings: List[str], keyword: str)`

Esta función recibe como argumentos una lista de cadenas y una cadena con una palabra a buscar, y devuelve la lista de los índices de las cadenas que contienen dicha palabra, independientemente de que haya mayúsculas o minúsculas. Nótese que hay que buscar palabras enteras, no partes de otras palabras. En la lista de cadenas puede haber caracteres no alfabéticos (símbolos) tras alguna palabra y eso no debe interferir con la búsqueda.

**Nota**: Para simplificar el código puedes asumir que los únicos caracteres no alfabéticos que aparecen en la lista de cadenas son el punto y la coma.

In [9]:
frases = [
    "El gato está en el jardín.",
    "Tengo una gata muy bonita,",
    "Gatear es bueno para los bebés.",
    "El gato, junto al perro.",
    "Mañana iré a comprar.",
    " Gato "
]

In [11]:
# Solución
import string
from typing import List
def word_search(strings: List[str], keyword: str):
    indices=[]
    for i in range(len(strings)):
        frase=strings[i]
        palabras=frase.replace(",", "").replace(".", "").split()
        for palabra in palabras:
            if palabra.lower()== keyword:
                indices.append(i+1)
                break
    return indices


print(word_search(frases, 'gato'))
print(word_search(frases, 'mañana'))
print(word_search(frases, 'perro'))

[1, 4, 6]
[5]
[4]


### Programa una función llamada `word_count(path: str)`

Esta función debe recibir como parámetro la ruta a un archivo de texto compuesto por diferentes líneas que contienen palabras, e imprimir el número de palabras que contiene el archivo. Si el archivo no existe, la función debe capturar la excepción e imprimir un mensaje de error. Por ejemplo, `word_count("palabras")` imprimiría `70` mientras que `word_count("nopalabras")` mostraría `"Archivo 'nopalabras' no encontrado."`. Nótese que en el archivo puede haber caracteres no alfabéticos (símbolos) entre algunas palabras que no deben interferir con el cálculo del resultado.

**Nota**: Para simplificar el código puedes asumir que los únicos caracteres no alfabéticos que aparecen en la lista de cadenas son el punto y la coma.

In [ ]:
# Solución

word_count("palabras")
word_count("nopalabras")

### Programa una función llamada `unique_words(path: str)`
 
Esta función debe recibir como parámetro la ruta a un archivo de texto compuesto por diferentes líneas que contienen palabras, y devolver una lista con todas las palabras diferentes que contiene el archivo ordenadas alfabéticamente (no tengas en cuenta que las tildes se ordenan incorrectamente), independientemente de que haya mayúsculas o minúsculas. Nótese que en la lista de cadenas puede haber caracteres no alfabéticos (símbolos) entre algunas palabras que deben ignorarse, es decir, no deben aparecer en la lista. Por ejemplo, `unique_words("palabras")` devolvería la lista `['a', 'alfabéticos', 'alguna', 'argumentos', ..., 'índices']`.

**Nota**: Para simplificar el código puedes asumir que los únicos caracteres no alfabéticos que aparecen en la lista de cadenas son el punto y la coma.

In [ ]:
# Solución

print(unique_words("palabras"))

### Programa una función llamada `letter_count(path: str)`

Esta función debe recibir como parámetro la ruta a un archivo de texto compuesto por diferentes líneas que contienen palabras, y devolver una lista de tuplas `(carácter, apariciones)` con las tuplas ordenadas alfabéticamente (no tengas en cuenta que las tildes se ordenan incorrectamente). Deben omitirse todos los caracteres no alfabéticos y los espacios. Además, no debe distinguirse entre mayúsculas minúsculas. Por ejemplo, `letter_count("palabras")` devolvería la lista `[('a', 57), ('b', 12), ('c', 19), ..., ('ú', 3)]`.

In [ ]:
# Solución

print(letter_count("palabras"))

# Ejercicios sobre Programación Orientada a Objectos

## Crea una jerarquía con las clases `Persona`, `Alumno` y `Profesor`.

Dada la siguiente definición de la clase `Persona`:

In [ ]:
class Persona(object):
    """ Una clase sencilla. """   # esto es un docstring
    especie = "Homo Sapiens"   # atributo de clase compartido entre todas las instancias
    
    def __init__(self, nombre):   # método mágico (dunder method)
        """ Este es el inicializador. """
        self.nombre = nombre   # atributo de instancia diferente para cada objeto
        
    def __str__(self):   # método mágico (dunder method)
        """ Este método lo usa Python cuando necesita imprimir el objeto como una cadena.
        Por ejemplo, lo usa la función print(). """
        return self.nombre

    def __repr__(self):   # método mágico (dunder method)
        """ Este método lo usa Python cuando necesita representar objeto como una cadena.
        La cadena resultante se puede ejecutar o evaluar. """
        return f"{__class__.__name__}({self.nombre})"

    def rename(self, nuevo_nombre):   # método común
        """ Reasigna el nombre y lo imprime. """
        self.nombre = nuevo_nombre
        print(f"Ahora me llamo {self.nombre}")


* Extiende la clase `Persona`:
  * El inicializador debe definir un atributo edad con un valor por defecto de 18 si no se especifica.
  * Crea un *getter* y un *setter* para una *property* edad. El *setter* debe comprobar que la edad es un entero mayor que o igual a 16 y menor que o igual a 120. Si no se cumple la condición, la asignación no se lleva a cabo.
* Escribe dos subclases de la clase Persona llamadas Alumno y Profesor.
  * La subclase `Alumno` debe definir los siguientes atributos y métodos:
    * Atributos:
      * `_notas: Dict[str, float]` (relaciona cada asignatura con su nota)
    * Propiedades:
      * `notas: str` (devuelve un diccionario `Dict[str, Calificacion]` de la forma "Ai: Ci" dónde `A` es la asignatura y `C` la calificación)
        * La calificación debe ser un enumerado con los siguientes valores: "NP" (`None`), "SUSPENSO" [0..5), "APROBADO" [5..6], "BIEN" [6..7), "NOTABLE" [7..9) y "SOBRESALIENTE" [9..10]
    * Métodos:
      * `matricula_asignatura(asignatura: str)`: añade `asignatura` a `_notas` con una nota inicial `None`.
        * Si el alumno ya está matriculado en `asignatura`, se mostrará el mensaje "El alumno '{nombre}' YA está matriculado en '{asignatura}'".
      * `elimina_asignatura(asignatura: str)`: elimina `asignatura` de `_notas`.
        * Si el alumno no está matriculado en `asignatura`, se mostrará el mensaje "El alumno '{nombre}' NO está matriculado en '{asignatura}'".
      * `califica_asignatura(asignatura: str, nota: float)`: asigna la calificación `nota` a la asignatura `asignatura`.
        * Si el alumno no está matriculado en `asignatura`, se mostrará el mensaje de error "El alumno '{nombre}' NO está matriculado en '{asignatura}'".
        * Si la nota no es un valor real entre 0.0 y 10.0, se emitirá una excepción `ValueError`.
      * `nota_asignatura(asignatura: str)`: devuelve la nota numérica de `asignatura`
        * Si el alumno no está matriculado en `asignatura`, se mostrará el mensaje "El alumno '{nombre}' NO está matriculado en '{asignatura}'" y se devolverá `None`.
        * Si la asignatura `asignatura` no tiene una nota numérica, se mostrará el mensaje "El alumno '{nombre}' NO ha sido calificado en '{asignatura}'" y se devolverá `None`.
      * `media_expediente()`: devuelve la media de las notas aprobadas (nota numérica mayor o igual a 5.0) del alumno.
        * Si el alumno no tiene ninguna asignatura aprobada, se devolverá `None`.
  * La subclase `Profesor` debe definir los siguientes atributos y métodos:
    * Atributos:
      * `_asignaturas: Set[str]` (**conjunto** de asignaturas impartidas por el profesor)
    * Propiedades:
      * `asignaturas: List[str]` (devuelve la lista de asignaturas impartidas por el profesor ordenadas alfabéticamente)
      * `num_asignaturas: int` (devuelve el número de asignaturas impartidas por el profesor)
    * Métodos:
      * `imparte_asignatura(asignatura: str)`: añade `asignatura` a `_asignaturas`.
        * Si el profesor ya imparte `asignatura`, se mostrará el mensaje "El profesor '{nombre}' YA imparte '{asignatura}'".
      * `elimina_asignatura(asignatura: str)`: elimina `asignatura` de `_asignaturas`.
        * Si el profesor no imparte `asignatura`, se mostrará el mensaje "El profesor '{nombre}' NO imparte '{asignatura}'".

In [ ]:
# Solución


In [ ]:
# Celda para testear las clases Persona, Alumno y Profesor
# Incluye aquí código de la funcionalidad pedida para comprobar su correcto funcionamiento:
